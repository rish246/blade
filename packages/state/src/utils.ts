// TODO: Later we can optimize it and create a treewalker type algorithm
export const deepEqualObject = (first: Object, second: Object): boolean => {
    return JSON.stringify(first) === JSON.stringify(second);
};
