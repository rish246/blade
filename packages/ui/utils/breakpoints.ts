export const breakpoints = {
    sm: 640,
    md: 768,
    lg: 1024,
    xl: 1280,
    "2xl": 1536,
};

export type Breakpoint = keyof typeof breakpoints;

export const mediaQuery = (key: keyof typeof breakpoints) =>
    `@media (min-width: ${breakpoints[key]}px)`;

export const getActiveBreakpoint = (): Breakpoint => {
    const width = window.innerWidth;
    const result = (() => {
        if (width >= breakpoints["2xl"]) return "2xl";
        if (width >= breakpoints.xl) return "xl";
        if (width >= breakpoints.lg) return "lg";
        if (width >= breakpoints.md) return "md";
        return "sm";
    })();

    return result;
};
