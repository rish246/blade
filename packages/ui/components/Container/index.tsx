import { useState, useEffect, useMemo, PropsWithChildren } from "react";
import { getActiveBreakpoint, type Breakpoint } from "../../utils/breakpoints";
import Box from "../Box";
export type ContainerProps = React.HTMLAttributes<HTMLDivElement> &
    PropsWithChildren<{
        maxWidth?:
            | "sm"
            | "md"
            | "lg"
            | "xl"
            | "2xl"
            | "full"
            | Partial<Record<Breakpoint, string>>;
        padding?: "none" | "sm" | "md" | "lg";
        /** Vertical margin */
        marginY?: "none" | "sm" | "md" | "lg";
        /** Centers the container horizontally */
        center?: boolean;
    }>;

export const Container: React.FC<ContainerProps> = ({
    children,
    maxWidth = "xl",
    padding = "md",
    marginY = "none",
    center = true,
    style,
    ...props
}) => {
    const [activeBp, setActiveBp] = useState<Breakpoint>("sm");

    useEffect(() => {
        if (!window) {
            return;
        }
        const handleResize = () => setActiveBp(getActiveBreakpoint());
        handleResize();
        window.addEventListener("resize", handleResize);
        return () => window.removeEventListener("resize", handleResize);
    }, []);

    // Determine which width to apply
    const resolvedMaxWidth =
        typeof maxWidth === "string" ? maxWidth : maxWidth[activeBp] || "full";

    const containerStyle: React.CSSProperties = useMemo(() => {
        const maxWidthMap: Record<string, string> = {
            sm: "640px",
            md: "768px",
            lg: "1024px",
            xl: "1280px",
            xxl: "1536px",
            full: "100%",
        };

        const paddingMap: Record<string, string> = {
            none: "0",
            sm: "8px",
            md: "16px",
            lg: "32px",
        };

        const marginYMap: Record<string, string> = {
            none: "0",
            sm: "8px",
            md: "16px",
            lg: "32px",
        };
        return {
            width: "100%",
            boxSizing: "border-box",
            maxWidth: maxWidthMap[resolvedMaxWidth],
            paddingLeft: paddingMap[padding],
            paddingRight: paddingMap[padding],
            marginTop: marginYMap[marginY],
            marginBottom: marginYMap[marginY],
            marginLeft: center ? "auto" : undefined,
            marginRight: center ? "auto" : undefined,
            ...style,
        };
    }, [resolvedMaxWidth, padding, marginY, center, style]);
    return (
        <Box style={containerStyle} {...props}>
            {children}
        </Box>
    );
};

export default Container;
