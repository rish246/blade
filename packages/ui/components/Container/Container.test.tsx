import React from "react";
import { render, screen, cleanup, waitFor } from "@testing-library/react";
import Container from ".";
import { vi } from "vitest";
import { breakpoints } from "../../utils/breakpoints";

// A simple store for the current screen width

const setScreenWidth = async (width: number) => {
    (window as any).innerWidth = width;
    window.dispatchEvent(new Event("resize"));
};

describe("Container Component", () => {
    beforeEach(async () => {
        cleanup();
        setScreenWidth(1024);
    });

    it("renders children properly", () => {
        render(<Container>Child content</Container>);
        expect(screen.getByText("Child content")).toBeInTheDocument();
    });

    it("applies default inline styles", () => {
        render(<Container>Default</Container>);
        const el = screen.getByText("Default");
        expect(el).toHaveStyle({
            width: "100%",
            maxWidth: "1280px",
            paddingLeft: "16px",
            paddingRight: "16px",
            marginLeft: "auto",
            marginRight: "auto",
        });
    });

    it("accepts custom maxWidth prop", () => {
        render(<Container maxWidth="md">Medium</Container>);
        const el = screen.getByText("Medium");
        expect(el).toHaveStyle({ maxWidth: "768px" });
    });

    it("disables centering when center=false", () => {
        render(<Container center={false}>Uncentered</Container>);
        const el = screen.getByText("Uncentered");
        expect(el.style.marginLeft).toBe("");
        expect(el.style.marginRight).toBe("");
    });

    it("applies padding variants correctly", () => {
        render(<Container padding="lg">Padding</Container>);
        const el = screen.getByText("Padding");
        expect(el).toHaveStyle({
            paddingLeft: "32px",
            paddingRight: "32px",
        });
    });

    it("applies margin variants correctly", () => {
        render(<Container marginY="lg">Margin</Container>);
        const el = screen.getByText("Margin");
        expect(el).toHaveStyle({
            marginTop: "32px",
            marginBottom: "32px",
        });
    });

    it("merges custom inline styles", () => {
        render(
            <Container
                style={{
                    backgroundColor: "red",
                    paddingLeft: "50px",
                }}
            >
                Styled
            </Container>,
        );
        const el = screen.getByText("Styled");
        expect(el.style.backgroundColor).toBe("red");
        expect(el).toHaveStyle({
            paddingLeft: "50px",
        });
    });

    describe("Responsive behavior", () => {
        it("updates maxWidth dynamically on resize", async () => {
            render(
                <Container
                    maxWidth={{ sm: "sm", md: "md", lg: "lg", xl: "xl" }}
                >
                    Responsive
                </Container>,
            );

            const el = screen.getByText("Responsive");

            // Small screen
            setScreenWidth(500);
            await waitFor(() => expect(el.style.maxWidth).toBe("640px"));

            // Medium screen
            setScreenWidth(800);
            await waitFor(() => expect(el.style.maxWidth).toBe("768px"));

            // Large screen
            setScreenWidth(1200);
            await waitFor(() => expect(el.style.maxWidth).toBe("1024px"));

            // Extra large screen
            setScreenWidth(1300);
            await waitFor(() => expect(el.style.maxWidth).toBe("1280px"));
        });

        it("falls back to full width when no matching breakpoint", () => {
            render(<Container maxWidth={{ sm: "sm" }}>Fallback</Container>);
            setScreenWidth(1600);
            const el = screen.getByText("Fallback");
            expect(el.style.maxWidth).toBe("100%");
        });
    });

    describe("Advanced scenarios", () => {
        it("updates inline styles when props change", () => {
            const { rerender } = render(
                <Container maxWidth="sm">Dynamic</Container>,
            );
            const el = screen.getByText("Dynamic");
            expect(el).toHaveStyle({ maxWidth: "640px" });

            rerender(<Container maxWidth="xl">Dynamic</Container>);
            expect(el).toHaveStyle({ maxWidth: "1280px" });
        });

        it("passes through additional HTML attributes", () => {
            render(
                <Container data-testid="container" role="region">
                    Accessible
                </Container>,
            );
            const el = screen.getByTestId("container");
            expect(el.getAttribute("role")).toBe("region");
        });

        it("handles multiple responsive breakpoints correctly", async () => {
            render(
                <Container
                    maxWidth={{ sm: "sm", md: "md", lg: "xl" }}
                    data-testid="multi"
                >
                    Multi
                </Container>,
            );
            const el = screen.getByText("Multi");

            setScreenWidth(breakpoints.sm);
            await waitFor(() => expect(el.style.maxWidth).toBe("640px"));

            setScreenWidth(breakpoints.lg);
            await waitFor(() => expect(el.style.maxWidth).toBe("1280px"));
        });

        it("memoizes styles and avoids recomputing unnecessarily", () => {
            const { rerender } = render(<Container>Memo</Container>);
            const el = screen.getByText("Memo");
            const initialStyle = el.getAttribute("style");

            rerender(<Container>Memo</Container>); // no prop change
            const rerenderedStyle = el.getAttribute("style");

            expect(initialStyle).toBe(rerenderedStyle);
        });

        it("cleans up resize listener on unmount", () => {
            const removeListener = vi.spyOn(window, "removeEventListener");
            const { unmount } = render(<Container>Cleanup</Container>);
            unmount();
            expect(removeListener).toHaveBeenCalledWith(
                "resize",
                expect.any(Function),
            );
        });
    });

    describe("Snapshot testing", () => {
        it("matches default render snapshot", () => {
            const { asFragment } = render(<Container>Snapshot</Container>);
            expect(asFragment()).toMatchSnapshot();
        });

        it("matches responsive config snapshot", () => {
            const { asFragment } = render(
                <Container maxWidth={{ sm: "sm", md: "xl" }}>
                    SnapshotResponsive
                </Container>,
            );
            expect(asFragment()).toMatchSnapshot();
        });
    });
});
