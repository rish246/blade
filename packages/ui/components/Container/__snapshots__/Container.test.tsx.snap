// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Container Component > Snapshot testing > matches default render snapshot 1`] = `
<DocumentFragment>
  <div
    style="width: 100%; box-sizing: border-box; max-width: 1280px; padding-left: 16px; padding-right: 16px; margin: 0px auto;"
  >
    Snapshot
  </div>
</DocumentFragment>
`;

exports[`Container Component > Snapshot testing > matches responsive config snapshot 1`] = `
<DocumentFragment>
  <div
    style="width: 100%; box-sizing: border-box; max-width: 100%; padding-left: 16px; padding-right: 16px; margin: 0px auto;"
  >
    SnapshotResponsive
  </div>
</DocumentFragment>
`;
